File: Benchmarks_FSharp/Core/Types/Rectangle.fs
namespace Benchmarks.Core.Types

open System.Runtime.CompilerServices

[<Struct>]
type Rectangle =
    val mutable X: int
    val mutable Y: int
    val mutable Width: int
    val mutable Height: int
    new(x, y, width, height) = { X = x; Y = y; Width = width; Height = height }

    member this.Left = this.X
    member this.Right = this.X + this.Width
    member this.Top = this.Y
    member this.Bottom = this.Y + this.Height

    static member Empty = Rectangle(0, 0, 0, 0)

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    member this.Intersects(other: Rectangle) =
        this.Left < other.Right && this.Right > other.Left &&
        this.Top < other.Bottom && this.Bottom > other.Top

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    member this.Contains(px: int, py: int) =
        px >= this.Left && px < this.Right && py >= this.Top && py < this.Bottom

File: Benchmarks/Core/Types/Rectangle.cs
using System.Runtime.CompilerServices;

namespace Benchmarks.Core.Types
{
    public struct Rectangle(int x, int y, int width, int height)
    {
        public int X { get; set; } = x;
        public int Y { get; set; } = y;
        public int Width { get; set; } = width;
        public int Height { get; set; } = height;

        public int Left => X;
        public int Right => X + Width;
        public int Top => Y;
        public int Bottom => Y + Height;

        public static readonly Rectangle Empty = new(0, 0, 0, 0);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool Intersects(Rectangle other)
        {
            return Left < other.Right && Right > other.Left &&
                   Top < other.Bottom && Bottom > other.Top;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool Contains(int px, int py)
        {
            return px >= Left && px < Right && py >= Top && py < Bottom;
        }
    }
}
